feat: add corner positions to tooltip

SethDonohue opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting. Someone may have pushed the same thing before!

Provide a summary of your changes in the title field above.
-->

# Pull Request

## 📖 Description
This adds corner positions to the tooltip to allow for more options.

New positions include:
- top-left
- top-right
- top-start
- top-end

- bottom-left
- bottom-right
- bottom-start
- bottom-end


Fixed Positions:
<img width="255" alt="Screen Shot 2022-01-05 at 2 26 46 PM" src="https://user-images.githubusercontent.com/31681651/148298706-6e0a83df-8c07-46fa-aba9-c5ebd34e87f1.png">

Start / End:
<img width="255" alt="Screen Shot 2022-01-05 at 2 27 18 PM" src="https://user-images.githubusercontent.com/31681651/148298711-8c9b09e7-ca90-43ee-a7f0-9c63044fb30e.png">

Start / End RTL:
<img width="255" alt="Screen Shot 2022-01-05 at 2 27 30 PM" src="https://user-images.githubusercontent.com/31681651/148298713-2765f4e9-57ab-4b26-80d4-fd7703dfdeb1.png">


<!---
Provide some background and a description of your work.
What problem does this change solve?
Is this a breaking change, chore, fix, feature, etc?
-->

### 🎫 Issues
closes #5483 
<!---
* List and link relevant issues here.
-->

## 👩‍💻 Reviewer Notes

Feedback: Are these location names good?

Please ensure that keyboard functionality, (toggling and Escape actions) work as expected.

<!---
Provide some notes for reviewers to help them provide targeted feedback and testing.

Do you recommend a smoke test for this PR? What steps should be followed?
Are there particular areas of the code the reviewer should focus on?
-->

## ✅ Checklist

### General

<!--- Review the list and put an x in the boxes that apply. -->

- [x] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

<!--- Review the list and put an x in the boxes that apply. -->
<!--- Remove this section if not applicable. -->

- [ ] I have added a new component
- [x] I have modified an existing component
- [x] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

scomea commented:
There's also a pr out that just needs owner approval to enable a center position, so someone's gonna have to do a merge. :-)

SethDonohue commented:
> There's also a pr out that just needs owner approval to enable a center position, so someone's gonna have to do a merge. :-)
> 
> https://github.com/microsoft/fast/pull/5314/files

Rebased and updated. All looks  good from my perspective.

robarbms commented on packages/web-components/fast-foundation/src/tooltip/tooltip.ts:
```diff
@@ -395,6 +395,94 @@ export class Tooltip extends FoundationElement {
                 break;
 
+            case TooltipPosition.topLeft:
+                this.verticalPositioningMode = "dynamic";
```
Should verticalPositioningMode for all of these be "locktodefault" like the other original positions above? Only the default seems to use dynamic positioning.

SethDonohue commented on packages/web-components/fast-foundation/src/tooltip/tooltip.ts:
```diff
@@ -395,6 +395,94 @@ export class Tooltip extends FoundationElement {
+                this.verticalDefaultPosition = "top";
+                this.horizontalDefaultPosition = "left";
+                this.horizontalInset = "false";
+                this.verticalInset = "false";
```
removed already defaulted values, thanks!

SethDonohue commented:
Are any code owners able to review and/or merge? No rush, just pinging before I forget!

EisenbergEffect commented on packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:
```diff
@@ -83,6 +83,27 @@ export const tooltipStyles: (
+        }
+
+        ${context.tagFor(AnchoredRegion)}.top.right .tooltip,
+        ${context.tagFor(AnchoredRegion)}.bottom.right .tooltip {
```
I think we should extract the `context.tagFor(AnchoredRegion)` to the top of the function so we only have to make that call once.

SethDonohue replied:
Extracted!

EisenbergEffect commented on packages/web-components/fast-foundation/src/tooltip/tooltip.spec.md:
```diff
@@ -42,6 +42,14 @@ _Attributes:_
     -   right
     -   start
     -   end
+    -   top-left
```
Thanks for fixing up the spec!

EisenbergEffect commented on packages/web-components/fast-foundation/src/tooltip/tooltip.ts:
```diff
@@ -395,6 +395,62 @@ export class Tooltip extends FoundationElement {
                 break;
 
+            case TooltipPosition.topLeft:
+                this.verticalPositioningMode = "locktodefault";
```
Seems like everything but the default case uses `locktodefault`. Can we factor that out and reduce duplication?

chrisdholt reviewed (approved):
@SethDonohue looks like you have one merge conflict to resolve. Thanks!

SethDonohue commented:
@scomea @chrisdholt @nicholasrice I can't bypass the vailed steps to merge this PR, can one of you do so?

SethDonohue commented:
> @scomea @chrisdholt @nicholasrice I can't bypass the failed steps to merge this PR, can one of you do so?

Hmm on second look, it looks like the api-report is still getting modified, I'll update that.

SethDonohue commented:
@scomea @chrisdholt @nicholasrice any idea whats up with this linux build error?

nicholasrice commented:
@awentzel re the code-climate failure.

awentzel commented:
> @awentzel re the code-climate failure.

Code Climate has an issue currently, I have a filed a ticket and waiting for a response.  If. the failing job is not "required" it can still be merged.  No ETA on solution.

awentzel commented:
> > @awentzel re the code-climate failure.
> 
> Code Climate has an issue currently, I have a filed a ticket and waiting for a response. If. the failing job is not "required" it can still be merged. No ETA on solution.

I found all tests passed on rebase, thus merged.
